File: apps/ios/Exponential/UI/Agent/AgentOptionsRow.swift
import ExpCore
import ExpUI
import SwiftUI

/// EXP-825: the launch options as ONE muted inline line under the composer
/// card (Danny's variant B): Device, Account, Model, a Plan switch, the Resume
/// switch inline while a worktree makes it offerable (EXP-481), and a `⋯` pill
/// for the rest (`AgentOptionsSheet`: Effort, Subagent model, Ultracode).
/// Every pill is a picker or a toggle — no disabled controls, the footer under
/// the row explains what cannot start.
///
/// EXP-872 folded the Agent pill INTO the Account one: the list is every login
/// the picked machine reports across agents, and picking one implies its agent.
/// EXP-993 dropped the Repository pill: a phone never picks a chat's repo — the
/// team's first repository is the anchor.
struct AgentOptionsRow: View {
    let model: AgentComposerModel

    @State private var showsMore = false

    private var launch: LaunchOptionsState { model.launch }

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                devicePill
                accountPill
                modelPill
                planPill
                resumePill
                morePill
            }
            .padding(.horizontal, 2)
        }
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("agent-options-row")
    }

    // MARK: - Pills

    /// The machine the run lands on — always named once one resolves (the
    /// desktop and web say it too); a picker only while there is a choice, a
    /// lone machine reads as a plain label like a lone agent does.
    @ViewBuilder
    private var devicePill: some View {
        if let device = model.device {
            if model.candidateDevices.count > 1 {
                // EXP-1030: the SHARED `DevicePicker`; the pill is only its
                // trigger. EXP-862 stands — a machine is recognised by its own
                // glyph (EXP-924: its owner's pick, else the kind default), and
                // the picker leads every row with it, exactly as the trigger
                // draws it. A shared machine names its owner under it. The
                // ONE row bridge (`DevicePickerDevice(SteerDevice)`) decides
                // all of that.
                DevicePicker(
                    devices: model.candidateDevices.map(DevicePickerDevice.init),
                    value: device.deviceId,
                    onChange: { model.selectDevice($0) },
                    trigger: {
                        OptionPillLabel(
                            icon: DeviceIconDisplay.iconName(for: device),
                            text: LaunchVocabulary.deviceName(device)
                        )
                    }
                )
                .accessibilityLabel("Device")
                .accessibilityIdentifier("agent-device-pill")
            } else {
                OptionPillLabel(
                    icon: DeviceIconDisplay.iconName(for: device),
                    text: LaunchVocabulary.deviceName(device),
                    chevron: false
                )
                .accessibilityLabel("Device")
                .accessibilityIdentifier("agent-device-pill")
            }
        }
    }

    /// EXP-872: the ONE account picker — the agent pill folded into it. Every
    /// signed-in login the picked machine reports, across agents, brand mark +
    /// email, the machine's default first; picking one picks its agent too.
    /// EXP-849: a login the agent REFUSED stays on offer wearing its badge, or
    /// a run starts and dies on an expired credential.
    ///
    /// EXP-1030: `AccountPickerMenu` is the trigger + the lone-login rule over
    /// the SHARED `AccountPicker` (`ExpUI`) — the pill opens the one picker
    /// sheet now, brand mark + email per row with the EXP-992 limit bars under
    /// each, instead of a menu of its own.
    ///
    /// EXP-642: the store slide's pop-out rect is measured off the agent
    /// control, so the identifier stays on this one.
    private var accountPill: some View {
        let options = launch.accountOptions(on: model.device)
        return AccountPickerMenu(
            options: options,
            selection: launch.selectedAccount(in: options),
            mark: { AgentBrandMark.image($0) },
            onSelect: { model.selectAccount($0) }
        )
        .accessibilityLabel("Account")
        .accessibilityIdentifier("start-coding-agent-picker")
    }

    private var modelPill: some View {
        GlassMenu {
            ForEach(LaunchVocabulary.modelValues(for: launch.agent), id: \.self) { value in
                GlassMenuItem(LaunchVocabulary.modelLabel(value)) {
                    launch.model = value
                }
            }
        } label: {
            OptionPillLabel(text: LaunchVocabulary.modelLabel(launch.model))
        }
        .accessibilityLabel("Model")
    }

    /// Plan mode is claude's (EXP-441/EXP-849); a resume never re-enters plan
    /// mode (the machine clamps it too), so the switch hides while one is on.
    /// EXP-827: a slide switch on every platform (web and desktop use one),
    /// not a lit select pill. EXP-859 rebuilt it without `.fixedSize()` and
    /// without scaling the app-wide toggle down: a scaled switch reported its
    /// UNSCALED size, which pushed the caption out of the pill and clipped the
    /// track. One caption, a plain toggle, both inside the capsule.
    @ViewBuilder
    private var planPill: some View {
        if LaunchVocabulary.supportsPlanMode(launch.agent), !model.resumeActive {
            @Bindable var launch = model.launch
            Toggle(isOn: $launch.planMode) {
                Text("Plan")
                    .font(.caption.weight(.medium))
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                    .lineLimit(1)
            }
            .toggleStyle(OptionPillToggleStyle())
            .accessibilityLabel("Plan mode")
        }
    }

    /// EXP-481: offered only while the picked machine's synced worktree
    /// inventory carries a row for the ONE checked issue.
    @ViewBuilder
    private var resumePill: some View {
        if model.resumeCandidate != nil {
            GlassPill(
                "Resume",
                mode: .select(isSelected: launch.resume) { launch.resume.toggle() }
            )
            .accessibilityLabel("Resume previous session")
        }
    }

    private var morePill: some View {
        Button {
            showsMore = true
        } label: {
            OptionPillLabel(icon: AppIcons.uiMore, text: "", chevron: false)
        }
        .buttonStyle(.plain)
        .accessibilityLabel("More options")
        .sheet(isPresented: $showsMore) {
            AgentOptionsSheet(model: model)
        }
    }
}

/// One option pill's LABEL — a glyph, a value and a chevron, on the row fill.
/// A label, not a button: `GlassMenu` and `GlassPicker` wrap it in a trigger
/// of their own, so a `GlassPill` (a `Button` itself) would nest two. The
/// agent's brand-marked trigger is `AgentPickerTriggerLabel` (ExpUI) since
/// EXP-862.
struct OptionPillLabel: View {
    var icon: String? = nil
    let text: String
    var chevron: Bool = true

    var body: some View {
        HStack(spacing: 5) {
            if let icon {
                AppIcon(icon, size: 12)
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
            }
            if !text.isEmpty {
                Text(text)
                    .font(.caption.weight(.medium))
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                    .lineLimit(1)
            }
            if chevron {
                AppIcon(AppIcons.uiChevronDown, size: 10)
                    .foregroundStyle(.white.opacity(TextOpacity.tertiary))
            }
        }
        .padding(.horizontal, text.isEmpty ? 8 : 10)
        .frame(height: 28)
        .background(GlassTokens.fillRow, in: Capsule())
        .overlay(Capsule().stroke(GlassTokens.strokeCard, lineWidth: GlassTokens.hairline))
        .contentShape(Capsule())
    }
}

/// EXP-825: the `⋯` sheet — the options that did not earn a pill: Effort
/// (Reasoning / Thinking per agent), the Subagent model (EXP-981, claude only)
/// and Ultracode (claude only — it IS `--effort ultracode`, so it disables the
/// Effort row). EXP-862 promoted the Account out of here into the row itself.
/// No MCP-server picker: mobile has none.
///
/// The Subagent model sits here rather than on the pill row: the row already
/// carries the pills a phone can hold, and this is the same place its sibling
/// Effort lives.
///
/// EXP-994: ONE grouped card, rows separated by `GlassDivider` hairlines —
/// the Settings idiom. It used to be a 2pt-gapped stack of individually
/// bordered rows, the shape every grouped list on this client stopped using.
struct AgentOptionsSheet: View {
    let model: AgentComposerModel

    var body: some View {
        @Bindable var launch = model.launch
        GlassSheetChrome(title: "Options") {
            VStack(spacing: 0) {
                GlassPickerRow(
                    LaunchVocabulary.effortTitle(for: launch.agent),
                    selection: $launch.effort,
                    options: [LaunchVocabulary.cliDefault] + LaunchVocabulary.effortValues(for: launch.agent),
                    label: { value in
                        value == LaunchVocabulary.cliDefault
                            ? "CLI default"
                            : LaunchVocabulary.effortLabel(value)
                    },
                    enabled: !(launch.agent == "claude" && launch.ultracode)
                )
                .padding(.horizontal, 12)
                .padding(.vertical, 12)

                // EXP-981: the model this run's SUBAGENTS get. Claude-only,
                // hidden for every other agent exactly like Ultracode.
                if LaunchVocabulary.supportsSubagentModel(launch.agent) {
                    GlassDivider()
                    GlassPickerRow(
                        "Subagent model",
                        selection: $launch.subagentModel,
                        options: LaunchVocabulary.subagentModelValues(),
                        label: { LaunchVocabulary.subagentModelLabel($0) }
                    )
                    .padding(.horizontal, 12)
                    .padding(.vertical, 12)
                    .accessibilityIdentifier("agent-subagent-model-row")
                }

                if launch.agent == "claude" {
                    GlassDivider()
                    Toggle("Ultracode", isOn: $launch.ultracode)
                        .tint(DesignTokens.Palette.primary)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 8)
                }
            }
            .glassSection()
            .padding(.horizontal, GlassSheetTokens.headerHPadding)
            .padding(.bottom, 16)
        }
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("agent-options-sheet")
    }
}

/// EXP-859: the options row's switch — the pill IS the toggle. The app-wide
/// glass switch is UISwitch-sized (51×31) and too tall for the 28pt row, and
/// scaling it down left SwiftUI laying out the unscaled size, which is what
/// clipped the track and shoved the caption out of the capsule. This draws the
/// track at the row's own scale instead, so nothing is transformed.
struct OptionPillToggleStyle: ToggleStyle {
    @Environment(\.motion) private var motion

    func makeBody(configuration: Configuration) -> some View {
        Button {
            withAnimation(motion.fast) { configuration.isOn.toggle() }
        } label: {
            HStack(spacing: 6) {
                configuration.label
                track(isOn: configuration.isOn)
            }
            .padding(.leading, 10)
            .padding(.trailing, 5)
            .frame(height: 28)
            .background(GlassTokens.fillRow, in: Capsule())
            .overlay(Capsule().stroke(GlassTokens.strokeCard, lineWidth: GlassTokens.hairline))
            .contentShape(Capsule())
        }
        .buttonStyle(.plain)
        .accessibilityAddTraits(.isToggle)
        .accessibilityValue(configuration.isOn ? "On" : "Off")
    }

    /// The glass switch at row scale: same colours and the same 2pt thumb
    /// inset, 34×20 instead of 51×31.
    private func track(isOn: Bool) -> some View {
        Capsule()
            .fill(isOn ? DesignTokens.Palette.primary : GlassTokens.fillCard)
            .overlay(
                Capsule().stroke(
                    isOn ? Color.clear : GlassTokens.strokeCard,
                    lineWidth: GlassTokens.hairline
                )
            )
            .overlay(alignment: isOn ? .trailing : .leading) {
                Circle()
                    .fill(
                        isOn
                            ? DesignTokens.Palette.primaryForeground
                            : DesignTokens.Palette.mutedForeground
                    )
                    .frame(width: 16, height: 16)
                    .padding(2)
            }
            .frame(width: 34, height: 20)
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/agent/AgentOptionsRow.kt
package com.exponential.app.ui.agent

import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.requiredSize
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.toggleableState
import androidx.compose.ui.state.ToggleableState
import androidx.compose.ui.unit.dp
import com.exponential.app.data.api.SteerDevice
import com.exponential.app.data.db.DeviceWorktreeEntity
import com.exponential.app.domain.AccountOption
import com.exponential.app.ui.components.AccountPickerPill
import com.exponential.app.ui.components.CLI_DEFAULT_EFFORT
import com.exponential.app.ui.components.DEFAULT_AGENT
import com.exponential.app.ui.components.GlassDropdownMenu
import com.exponential.app.ui.components.GlassMenuItem
import com.exponential.app.ui.components.GlassPill
import com.exponential.app.ui.components.GlassSheet
import com.exponential.app.ui.components.OptionGroup
import com.exponential.app.ui.components.GroupDivider
import com.exponential.app.ui.components.PickerRow
import com.exponential.app.ui.components.PillMode
import com.exponential.app.ui.components.SwitchRow
import com.exponential.app.ui.components.SwitchThumb
import com.exponential.app.ui.components.deviceIcon
import com.exponential.app.ui.components.deviceOptionLabel
import com.exponential.app.ui.components.effortLabel
import com.exponential.app.ui.components.effortValuesFor
import com.exponential.app.ui.components.glassSwitchColors
import com.exponential.app.ui.components.modelLabel
import com.exponential.app.ui.components.modelOptionsFor
import com.exponential.app.ui.components.SUBAGENT_MODEL_LABEL
import com.exponential.app.ui.components.subagentModelLabel
import com.exponential.app.ui.components.subagentModelOptions
import com.exponential.app.ui.components.supportsPlanMode
import com.exponential.app.ui.components.supportsSubagentModel
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.theme.TextEmphasis

/**
 * EXP-825: the launch options as ONE muted inline line under the composer
 * card (Danny's variant B): Device, Account, Model, a Plan switch, the Resume
 * switch inline while a worktree makes it offerable (EXP-481), and a `⋯` pill
 * for the rest ([AgentOptionsSheet]: Effort, Subagent model, Ultracode).
 * Every pill is a menu or a toggle — no disabled controls; the caption under
 * the row explains what cannot start. Horizontally scrolling: a phone cannot
 * fit six pills, and wrapping would push the sessions list around.
 *
 * EXP-872: there is no agent pill any more — the ACCOUNT picker carries the
 * brand mark, and picking a login implies its agent. EXP-993: and no
 * repository pill: mobile never showed the chat's optional anchor well, so a
 * chat simply takes the team's first repository.
 */
@Composable
internal fun AgentOptionsRow(
    devices: List<SteerDevice>,
    device: SteerDevice?,
    onDeviceChange: (String) -> Unit,
    launch: LaunchDraft,
    /** EXP-872: the settled machine's logins, device default first. */
    accountOptions: List<AccountOption>,
    /** A picked login sets the agent AND the account in one go. */
    onAccountChange: (AccountOption) -> Unit,
    onModelChange: (String) -> Unit,
    onPlanModeChange: (Boolean) -> Unit,
    resumeCandidate: DeviceWorktreeEntity?,
    resume: Boolean,
    onResumeChange: (Boolean) -> Unit,
    onMore: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // A resume never re-enters plan mode (EXP-202) — the switch hides while
    // one is on, like the machine's own clamp.
    val resumeActive = resume && resumeCandidate != null
    Row(
        modifier = modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState())
            .testTag("agent-options-row"),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        // The machine the run lands on — always named once one resolves (the
        // desktop and web say it too); a menu only while there is a choice,
        // a lone machine reads as a plain label like a lone agent does.
        if (device != null) {
            OptionMenuPill(
                icon = deviceIcon(device),
                text = deviceOptionLabel(device),
                contentDescription = "Device",
                options = devices.map { it.deviceId },
                optionLabel = { id -> devices.firstOrNull { it.deviceId == id }?.let(::deviceOptionLabel) ?: id },
                // EXP-862: a picker whose VALUE carries a glyph carries it on
                // the items too — the machine's own icon, here (EXP-924: its
                // owner's pick, else the kind default).
                optionIcon = { id ->
                    val row = devices.firstOrNull { it.deviceId == id }
                    deviceIcon(row?.icon, row?.isServer == true)
                },
                selected = device.deviceId,
                onSelect = onDeviceChange,
                enabled = devices.size > 1,
                modifier = Modifier.testTag("agent-device-pill"),
            )
        }
        // EXP-872: the ACCOUNT — the ONE picker every launch surface renders.
        // The brand mark plus the login's email; picking one implies its
        // agent, so there is no agent pill beside it.
        // EXP-642: the store slide's pop-out rect used to be measured off the
        // segmented strip this replaced, so the testTag stays on this pill.
        AccountPickerPill(
            options = accountOptions,
            selectedKey = launch.accountKey,
            onSelect = onAccountChange,
            modifier = Modifier.testTag("start-coding-agent-picker"),
        )
        OptionMenuPill(
            text = modelLabel(launch.model),
            contentDescription = "Model",
            options = modelOptionsFor(launch.agent),
            optionLabel = ::modelLabel,
            selected = launch.model,
            onSelect = onModelChange,
        )
        // Plan mode is claude-only since EXP-849. EXP-827: a slide switch on
        // every platform, not a lit select pill.
        if (supportsPlanMode(launch.agent) && !resumeActive) {
            PlanSwitchPill(
                checked = launch.planMode,
                onCheckedChange = onPlanModeChange,
            )
        }
        // EXP-481: offered only while the picked machine's synced worktree
        // inventory carries a row for the ONE checked issue.
        if (resumeCandidate != null) {
            GlassPill(
                "Resume",
                onClick = { onResumeChange(!resume) },
                mode = PillMode.Select,
                selected = resume,
                contentDescription = "Resume previous session",
            )
        }
        GlassPill(
            "",
            icon = ExpIcons.uiMore,
            onClick = onMore,
            contentDescription = "More options",
        )
    }
}

/**
 * EXP-827: the Plan switch on the row's capsule. The whole pill is the tap
 * target and reads as ONE switch to TalkBack; the M3 switch inside is a pure
 * indicator ([glassSwitchColors] chrome), scaled to sit in the 28dp capsule.
 */
@Composable
private fun PlanSwitchPill(
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
) {
    GlassPill(
        "Plan",
        onClick = { onCheckedChange(!checked) },
        trailing = {
            Box(
                modifier = Modifier.size(width = 36.dp, height = 22.dp),
                contentAlignment = Alignment.Center,
            ) {
                Switch(
                    checked = checked,
                    onCheckedChange = null,
                    colors = glassSwitchColors(),
                    thumbContent = SwitchThumb,
                    modifier = Modifier
                        .requiredSize(width = 52.dp, height = 32.dp)
                        .scale(0.7f),
                )
            }
        },
        contentDescription = "Plan mode",
        modifier = Modifier.semantics {
            role = Role.Switch
            toggleableState = if (checked) ToggleableState.On else ToggleableState.Off
        },
    )
}

/**
 * One option pill that opens a menu of its values: a glyph, the picked value
 * and a chevron, on the capsule every other pill wears. [enabled] false
 * renders the label alone — a single value is a statement, not a choice.
 *
 * EXP-862: [optionIcon] puts the value's own glyph on the MENU rows too — a
 * picker whose trigger shows an icon shows it on its items.
 */
@Composable
private fun OptionMenuPill(
    text: String,
    contentDescription: String,
    options: List<String>,
    optionLabel: (String) -> String,
    selected: String?,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    optionIcon: ((String) -> ImageVector?)? = null,
    enabled: Boolean = true,
) {
    var open by remember { mutableStateOf(false) }
    Box(modifier = modifier) {
        GlassPill(
            text,
            onClick = if (enabled) ({ open = true }) else null,
            icon = icon,
            trailing = if (enabled) {
                {
                    Icon(
                        ExpIcons.uiChevronDown,
                        contentDescription = null,
                        modifier = Modifier.size(10.dp),
                        tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                    )
                }
            } else {
                null
            },
            contentDescription = contentDescription,
        )
        GlassDropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            options.forEach { option ->
                val glyph = optionIcon?.invoke(option)
                GlassMenuItem(
                    text = { Text(optionLabel(option)) },
                    leadingIcon = when {
                        glyph != null -> {
                            { Icon(glyph, contentDescription = null, modifier = Modifier.size(16.dp)) }
                        }
                        option == selected -> {
                            { Icon(ExpIcons.uiCheck, contentDescription = null, modifier = Modifier.size(16.dp)) }
                        }
                        else -> null
                    },
                    // With a glyph in the leading slot the pick is marked at
                    // the trailing edge instead, so both can show at once.
                    trailingIcon = if (glyph != null && option == selected) {
                        { Icon(ExpIcons.uiCheck, contentDescription = null, modifier = Modifier.size(16.dp)) }
                    } else {
                        null
                    },
                    onClick = {
                        open = false
                        onSelect(option)
                    },
                )
            }
        }
    }
}

/**
 * EXP-825: the `⋯` sheet — the options that did not earn a pill: Effort
 * (Reasoning / Thinking per agent), the Subagent model (EXP-981, claude only)
 * and Ultracode (claude only — it IS `--effort ultracode`, so it disables the
 * Effort row). No MCP-server picker: mobile has none. EXP-862: the Account
 * moved OUT of here onto the options row itself — which login a run spends is
 * a decision, not an overflow entry.
 */
@Composable
internal fun AgentOptionsSheet(
    launch: LaunchDraft,
    onEffortChange: (String) -> Unit,
    onSubagentModelChange: (String) -> Unit,
    onUltracodeChange: (Boolean) -> Unit,
    onDismiss: () -> Unit,
) {
    GlassSheet(
        title = "Options",
        onDismiss = onDismiss,
        modifier = Modifier.testTag("agent-options-sheet"),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState()),
        ) {
            OptionGroup {
                PickerRow(
                    label = when (launch.agent) {
                        "codex" -> "Reasoning"
                        else -> "Effort"
                    },
                    value = effortLabel(launch.effort),
                    options = listOf(CLI_DEFAULT_EFFORT) + effortValuesFor(launch.agent),
                    selected = launch.effort,
                    optionLabel = ::effortLabel,
                    // Ultracode IS `--effort ultracode` — it owns the row.
                    enabled = !(launch.agent == DEFAULT_AGENT && launch.ultracode),
                    onSelect = onEffortChange,
                )
                // EXP-981: the model this run's SUBAGENTS spend, right beside
                // the run's own. claude only — it is the one agent that
                // spawns them — so it hides like Ultracode does.
                if (supportsSubagentModel(launch.agent)) {
                    GroupDivider()
                    PickerRow(
                        label = SUBAGENT_MODEL_LABEL,
                        value = subagentModelLabel(launch.subagentModel),
                        options = subagentModelOptions(),
                        selected = launch.subagentModel,
                        optionLabel = ::subagentModelLabel,
                        onSelect = onSubagentModelChange,
                    )
                }
                if (launch.agent == DEFAULT_AGENT) {
                    GroupDivider()
                    SwitchRow(
                        title = "Ultracode",
                        checked = launch.ultracode,
                        onCheckedChange = onUltracodeChange,
                    )
                }
            }
            Spacer(Modifier.height(16.dp))
        }
    }
}
